### PR TITLE
Moves the departmental delivery area check to a unit test so it stops spamming logs

### PR DESCRIPTION
**Pull request by Maia-J**
## About The Pull Request

So, thanks to the map not being loaded yet when jobs are initialized, the logs are needlessly spammed by a check that can never pass.

image

Also adds some possible locations to engineering and science

As such, I just moved all this logging stuff and screaming at mappers/coders into a unit test. I honestly only have very vague understanding of how these work so someone with more knowledge please check if I did everything right.

## Why It's Good For The Game

Cleaner logs and one less headache for people trying to find actual issues in logs.

🆑 
code: Improvements to handling how departmental orders are checked
\:cl:

### CODE REVIEW

File: code/modules/jobs/departments/departments.dm [hunk 1]
```diff
@@ -28,14 +28,15 @@
...
-	//every area fallback didn't exist on this map so throw a mapping error and set some generic area that uuuh please exist okay
-	log_mapping("[type] has no valid areas to deliver to on this map, add some more fallback areas to its \"department_delivery_areas\" var.")
-	department_delivery_areas = list(/area/station/hallway/primary/central) //if this doesn't exist like honestly fuck your map man
+	// Speeds up any future loops
+	if(check_areas == TRUE)
+		for(var/delivery_area_type in department_delivery_areas)
+			if(GLOB.areas_by_type[delivery_area_type])
+				department_delivery_areas = list(delivery_area_type)
+				return
+		department_delivery_areas = list(/area/station/hallway/primary/central) //if this doesn't exist like honestly fuck your map man
```
Comment (MrMelbert) [inline code/modules/jobs/departments/departments.dm:39]: We don't even need any code in `/datum/job_department/New` anymore, the unit test can handle this entirely

File: code/modules/unit_tests/cargo_dep_order_locations.dm [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+/datum/unit_test/cargo_dep_order_locations/Run()
+	for(var/datum/job_department/department as anything in subtypesof(/datum/job_department))
+		var/datum/job_department/dep = new department(FALSE)
+		var/delivery_areas = dep.department_delivery_areas
+		if(isnull(initial(dep.department_delivery_areas)) || !length(delivery_areas))
+			continue
+		if(check_valid_delivery_location(delivery_areas))
+			continue
+		else
+			TEST_FAIL("[department.type] failed to find a valid delivery location on this map.")
```
Comment (MrMelbert) [inline code/modules/unit_tests/cargo_dep_order_locations.dm:12]: And if we move all the code to this unit test, we don't need to `new` the departments, we can just check `SSjobs`

**Review by MrMelbert — CHANGES_REQUESTED** [review]
Good catch, fixing my mess-up

File: code/modules/unit_tests/cargo_dep_order_locations.dm (hunk 2 above)
Comment (Majkl-J) [reply code/modules/unit_tests/cargo_dep_order_locations.dm:12]: Valid points, on it

File: code/modules/unit_tests/cargo_dep_order_locations.dm [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+/datum/unit_test/cargo_dep_order_locations/Run()
+	for(var/datum/job_department/department as anything in SSjob.joinable_departments)
+		var/delivery_areas = department.department_delivery_areas
+		if(isnull(delivery_areas) || !length(delivery_areas))
+			continue
+		if(check_valid_delivery_location(delivery_areas))
+			continue
+		else
+			TEST_FAIL("[department.type] failed to find a valid delivery location on this map.")
```
Comment (MrMelbert) [inline code/modules/unit_tests/cargo_dep_order_locations.dm:11]:
```suggestion
		TEST_FAIL("[department.type] failed to find a valid delivery location on this map.")
```

File: code/modules/unit_tests/cargo_dep_order_locations.dm [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/datum/unit_test/cargo_dep_order_locations
+
+/datum/unit_test/cargo_dep_order_locations/Run()
+	for(var/datum/job_department/department as anything in SSjob.joinable_departments)
+		var/delivery_areas = department.department_delivery_areas
+		if(isnull(delivery_areas) || !length(delivery_areas))
+			continue
```
Comment (MrMelbert) [inline code/modules/unit_tests/cargo_dep_order_locations.dm:7]:
```suggestion
		if(!length(delivery_areas))
			continue
```
`!length(null)` is valid

**Review by MrMelbert — APPROVED** [review]

File: code/modules/jobs/departments/departments.dm [hunk 5]
```diff
@@ -144,7 +138,11 @@
 	label_class = "science"
 	ui_color = "#c973c9"
 	nation_prefixes = list("Sci", "Griffa", "Geneti", "Explosi", "Mecha", "Xeno", "Nani", "Cyto")
-	department_delivery_areas = list(/area/station/science/research)
+	department_delivery_areas = list(
+		/area/station/science/research,
+		/area/station/science/lobby,
+		/area/station/science/lab,
+		)
```
Comment (MrMelbert) [inline code/modules/jobs/departments/departments.dm:145]:
```suggestion
	)
```

**Pull request merged**